### PR TITLE
feat: improve build performance with Docker `--cache-from` flag

### LINKED ISSUE #4: Improve performance with --cache-from
If there was an option to use docker build's --cache-from here:
https://github.com/CircleCI-Public/aws-ecr-orb/blob/master/src/orb.yml#L73
It would presumably cut build time significantly in the same way Circle's proprietary layer caching does, but more reliably

Reference:
https://semaphoreci.com/docs/docker/docker-layer-caching.html
https://docs.docker.com/edge/engine/reference/commandline/build/

### PR DESCRIPTION
**Pull request by RanadeepPolavarapu**
### Checklist

- [X] All new jobs, commands, executors, parameters have descriptions
- [X] Examples have been added for any significant new features
- [X] README has been updated, if necessary

### Motivation, issues

Allows to achieve a performance boost.

### Description

Added `--cache-from` flag to the container build process.

Resolves #4 .

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -12,6 +12,7 @@ Following is the full list of parameters required by this orb's various commands
 | `create-repo` | `boolean` | `false` | set this to true to create the repository if it does not already exist |
 | `dockerfile` | `string` | `Dockerfile` | name of Dockerfile to use |
 | `executor` | `executor` | `default` | name of any custom executor (default is `machine: true`) |
+| `cache-from` | `string` | `""` | `docker build --cache-from` flag to speed up builds. Specify multiple image/tag combinations as a comma separated values. Example, `node:13,python:latest,openjdk:12,gcr.io/google-containers/busybox`. Please see, https://docs.docker.com/engine/reference/commandline/build/ |
```
Comment (kaikuchn) [inline README.md]:
```suggestion
| `cache-from` | `string` | `""` | `docker build --cache-from` flag to speed up builds. Specify multiple image/tag combinations as a comma separated list of values. For example, `node:13,python:latest,openjdk:12,gcr.io/google-containers/busybox`. Please see https://docs.docker.com/engine/reference/commandline/build/ |
```

File: src/commands/build-and-push-image.yml [hunk 2]
```diff
@@ -98,6 +98,15 @@ parameters:
     type: string
     default: .
 
+  cache-from:
+    type: string
+    default: ""
+    description: >
+      Images to consider as cache sources. Specified as comma separated values of
+      image and tag syntax.
+      Example, node:13,python:latest,openjdk:12,gcr.io/google-containers/busybox.
```
Comment (kaikuchn) [inline src/commands/build-and-push-image.yml]:
```suggestion
      For example, node:13,python:latest,openjdk:12,gcr.io/google-containers/busybox.
```

File: src/commands/build-and-push-image.yml [hunk 3]
```diff
@@ -98,6 +98,15 @@ parameters:
...
+      Please see, https://docs.docker.com/engine/reference/commandline/build/
```
Comment (kaikuchn) [inline src/commands/build-and-push-image.yml]:
```suggestion
      Please see https://docs.docker.com/engine/reference/commandline/build/
```

File: src/commands/build-image.yml [hunk 4]
```diff
@@ -32,6 +32,15 @@ parameters:
     type: string
     default: 10m
 
+  cache-from:
+    type: string
+    default: ""
+    description: >
+      Images to consider as cache sources. Specified as comma separated values of
+      image and tag syntax.
+      Example, node:13,python:latest,openjdk:12,gcr.io/google-containers/busybox.
```
Comment (kaikuchn) [inline src/commands/build-image.yml]:
```suggestion
      For example, node:13,python:latest,openjdk:12,gcr.io/google-containers/busybox.
```

File: src/commands/build-image.yml [hunk 5]
```diff
@@ -32,6 +32,15 @@ parameters:
...
+      Please see, https://docs.docker.com/engine/reference/commandline/build/
```
Comment (kaikuchn) [inline src/commands/build-image.yml]:
```suggestion
      Please see https://docs.docker.com/engine/reference/commandline/build/
```

File: src/jobs/build-and-push-image.yml [hunk 6]
```diff
@@ -100,6 +100,15 @@ parameters:
     type: string
     default: .
 
+  cache-from:
+    type: string
+    default: ""
+    description: >
+      Images to consider as cache sources. Specified as comma separated values of
+      image and tag syntax.
+      Example, node:13,python:latest,openjdk:12,gcr.io/google-containers/busybox.
```
Comment (kaikuchn) [inline src/jobs/build-and-push-image.yml]:
```suggestion
      For example, node:13,python:latest,openjdk:12,gcr.io/google-containers/busybox.
```

File: src/jobs/build-and-push-image.yml [hunk 7]
```diff
@@ -100,6 +100,15 @@ parameters:
...
+      Please see, https://docs.docker.com/engine/reference/commandline/build/
```
Comment (kaikuchn) [inline src/jobs/build-and-push-image.yml]:
```suggestion
      Please see https://docs.docker.com/engine/reference/commandline/build/
```

**Comment by RanadeepPolavarapu** [comment]
@kaikuchn Thanks for the feedback.

Added an extra check where the `--cache-from` flag will default to Docker's implicit build cache if the parameter's default value is specified (ex. if `cache-from` param is omitted).

**Comment by RanadeepPolavarapu** [comment]
Also, added an implicit `docker pull` from the comma separated values in `cache-from` param to avoid having to do an explicit `docker pull`. Please let me know if you would prefer not to have this.

**Comment by kaikuchn** [comment]
> Also, added an implicit `docker pull` from the comma separated values in `cache-from` param to avoid having to do an explicit `docker pull`. Please let me know if you would prefer not to have this.

Sounds reasonable to me, when people specify a cache-from image they probably want to use it after all 😄.
However, would this fail if an image cannot be pulled? I think it would be nicer to issue a warning in such a case, WDYT?

**Comment by RanadeepPolavarapu** [comment]
@kaikuchn Great point 😄 . I added a fail safe to notify user that the pull has failed and has been ignored and continue with the rest.

**Comment by KyleTryon** [comment]
Hello all. Thank you for your contribution but we are going to go with a solution posted here: https://github.com/CircleCI-Public/aws-ecr-orb/issues/4. Please feel free to submit changes to this PR if desired. Your contribution will still count for hacktoberfest and orbtoberfest. Much appreciated!

**Pull request closed without merging**